### ISSUE TITLE
Catastrophic performances on OCaml files

### ISSUE THREAD

**Issue by Niols** [tweag/topiary#519]
**Describe the bug**

On trivial OCaml files, Topiary in its latest version takes forever to run.

**To Reproduce**

On my machine, this takes about 20s:
```console
$ echo 'let () = ()' > test.ml
$ time topiary -if test.ml
```
It was confirmed on @aspiwack's machine as well.

**Expected behavior**

I would expect Topiary to run under one second.

**Environment**
 - OS name + version: NixOS
 - Version of the code: ~v0.2.1,~ v0.2.2 and master all seem affected.
     _edit: v0.2.1 is slow, but not catastrophically slow; cf https://github.com/tweag/topiary/issues/525_

**Additional context**

Bisect and flamegraph coming soon.

**Comment by ErinvanderVeen** [comment]
Attached is the flamegraph applied on the example OCaml in the issue above.
![flamegraph](https://github.com/tweag/topiary/assets/10973664/a01ec859-f841-44ce-b754-98453c90b4eb)

**Comment by ErinvanderVeen** [comment]
Nearly all of the time is spend in initializing the query structure in the tree-sitter crate. I don't think our OCaml queries are so crazy that they would cause such an issue, but we should investigate anyway. Otherwise this is an upstream issue that we've run into.

**Comment by Niols** [comment]
For a tiny bit of context, initially, I was just complaining to @nbacquey that Topiary was taking more than 1s per file for simple files. He then noticed that I was using v0.2.1 and asked me to try the latest version... that happened to be much much much worse. After we fix this, we should still investigate was to make normal runs faster. cf issue by @aspiwack above.

**Comment by aspiwack** [comment]
Was this closed by mistake? Or is there another tracking issue now that #522 is merged?

**Comment by Niols** [comment]
I think I can now open the issue I had initially about Topiary being quite slow (with quite slow = one second per file).

**Comment by aspiwack** [comment]
You can (though there is #523 which tracks some of it). But we still need to understand what's going on in the last version Ocaml grammar and query, and address this issue.

**Comment by ErinvanderVeen** [comment]
This was closed automatically by GitHub.

**Comment by Niols** [comment]
Quick benchmark of running Topiary (as of 738a178e2b728957c0e8eab85c03717923230c60) compiled in release mode on all the `*.ml` files of [Morsmall](https://github.com/colis-anr/morsmall). This consists in running:
```
time find -name '*.ml' -print -exec topiary -if {} \;
```
a few times while bisecting on [tree-sitter/tree-sitter-ocaml](https://github.com/tree-sitter/tree-sitter-ocaml/). I was expecting one commit changing the behaviour violently, but I actually found two so there were two different bisections. Here are the results:

| Commit                                                                       | Query | 1    | 2    | Time |
|------------------------------------------------------------------------------|-------|------|------|------|
| https://github.com/tree-sitter/tree-sitter-ocaml/commit/de07323 (`v0.20.1`)  | old   | good | good | 0:10 |
| https://github.com/tree-sitter/tree-sitter-ocaml/commit/f1106bf              | old   | good |      | 0:10 |
| https://github.com/tree-sitter/tree-sitter-ocaml/commit/d025214              | old   | bad  |      | 0:30 |
| https://github.com/tree-sitter/tree-sitter-ocaml/commit/cece53c              | old   | bad  |      | 0:30 |
| https://github.com/tree-sitter/tree-sitter-ocaml/commit/a09c63f              | old   | bad  | good | 0:30 |
| https://github.com/tree-sitter/tree-sitter-ocaml/commit/4823fe7              | old   |      | good | 0:30 |
| https://github.com/tree-sitter/tree-sitter-ocaml/commit/f4214a1              | new   |      | bad  | 1:45 |
| https://github.com/tree-sitter/tree-sitter-ocaml/commit/9724dfa  (`v0.20.2`) | new   |      | bad  | 2:05 |
| https://github.com/tree-sitter/tree-sitter-ocaml/commit/2da4930  (`master`)  | new   | bad  | bad  | 1:30 |

The column “query” tells whether I use Topiary as of 738a178e2b728957c0e8eab85c03717923230c60 or if I include the new query as per 738a178e2b728957c0e8eab85c03717923230c60. The columns 1 and 2 represent the two bisections and what I told git. The column “time” is to be taken with a grain of salt, especially for the longer times that I repeated less often. There seems to be two different performance drops making Topiary approximately 3x slower (which, as the flamegraph suggests, should be almost linear in the time spent in the query initialisation).

- https://github.com/tree-sitter/tree-sitter-ocaml/commit/d025214
- https://github.com/tree-sitter/tree-sitter-ocaml/commit/f4214a1 (although it is to be noted that our OCaml query has to be updated to follow this change).

**Comment by aspiwack** [comment]
The first change (https://github.com/tree-sitter/tree-sitter-ocaml/commit/d025214) merely adds the ability to write `()` as a functor argument (it makes the functor generative). How could such a small change possibly affect our query's build time?

**Comment by Niols** [comment]
The changes are massive, though, so I guess it all depends on what tools were used to re-generate the parsers from the grammar. Maybe those tools too were updated?

**Comment by Niols** [comment]
(It is also entirely possible that my measurements are flawed in one way or another.)

**Comment by ErinvanderVeen** [comment]
Between those two commits, where the time blows up, the number of internal tree-sitter state machine states increase from 9769 to 24361. So it is probably worth for the Topiary team to invest some time into optimizing the ocaml grammar.

**Comment by 314eter** [comment]
It looks like I was using tree-sitter-cli 0.20.7 to generate the parser before tree-sitter/tree-sitter-ocaml@d025214, and accidentally switched back to 0.20.6. That can explain the difference, since the actual change in the grammar is very small.

The number of states has indeed increased a lot because of tree-sitter/tree-sitter-ocaml@f4214a1. The actual structure of the grammar didn't change that much, but a lot of anonymous nodes became names nodes. It can be combination of slower parsing (although more states doesn't necessarily mean slower parsing) and the more complicated query. Operators are used a lot, so if the query becomes a bit slower that can have a big impact.

**Comment by aspiwack** [comment]
@314eter for the record, in our current measurements, we basically can't see the time spent parsing. The time is overwhelmingly dominated by _compiling_ the query. Compiling queries maybe scales super-linearly with the number of states of the grammar? This is what our current state of understanding would suggest.

**Comment by ErinvanderVeen** [comment]
```
cargo bench with tree-sitter 0.20.8:
    time:   [589.75 ms 592.02 ms 594.39 ms]

cargo bench with tree-sitter 0.20.6:
    time:   [4.6863 s 4.7138 s 4.7421 s]
```

**Comment by Niols** [comment]
@ErinvanderVeen Now that #533 is merged, is there still a reason to keep this one open?

**Comment by ErinvanderVeen** [comment]
@Niols Thank you, closed in favour of #539
